robot_modeling/DQ_SerialManipulator.m

```diff
@@ -67,32 +62,59 @@
         handle
         n_links;
     end
+
+    methods (Abstract)
+         % RAW_POSE_JACOBIAN(q) returns the Jacobian that satisfies 
+         % vec(x_dot) = J * q_dot, where x = fkm(q) and q is the 
+         % vector of joint variables.
+         %
+         % RAW_POSE_JACOBIAN(q,ith) returns the Jacobian that
+         % satisfies vec(x_ith_dot) = J * q_dot(1:ith), where 
+         % x_ith = fkm(q, ith), that is, the fkm up to the i-th link.
+         %
+         % This function does not take into account any base or
+         % end-effector displacements and should be used mostly
+         % internally in DQ_kinematics
+         J = raw_pose_jacobian(obj, q,to_ith_link);
+
+
+         % RAW_POSE_JACOBIAN_DERIVATIVE(q,q_dot) returns the Jacobian 
+         % time derivative.
+         % 
+         % RAW_POSE_JACOBIAN_DERIVATIVE(q,q_dot, to_ith_link) returns the first
+         % to_ith_link columns of the Jacobian time derivative.
+         % This function does not take into account any base or
+         % end-effector displacements.
+         J_dot = raw_pose_jacobian_derivative(obj, q, q_dot, to_ith_link);
+       
+         %   RAW_FKM(q) calculates the forward kinematic model and
+         %   returns the dual quaternion corresponding to the
+         %   last joint (the displacements due to the base and the effector
+         %   are not taken into account).
+         %
+         %   'q' is the vector of joint variables
+         %   'to_ith_link' defines until which link the raw_fkm will be
+         %   calculated.
+         %
+         %   This is an auxiliary function to be used mainly with the
+         %   Jacobian function.   
+         pose = raw_fkm(obj,q, to_ith_link); 
+    end
     
     methods
-        function obj = DQ_SerialManipulator(A,convention)
+        function obj = DQ_SerialManipulator(dim_configuration_space)
```

Comment:
What's the rationale behind including the `dim_configuration_space` parameter in the constructor of this abstract class?

Reply:
Hi @bvadorno, 

The implementation is based on the current version of C++ ([DQ_SerialManipulator.cpp] (https://github.com/dqrobotics/cpp/blob/master/src/robot_modeling/DQ_SerialManipulator.cpp)) implemented by @mmmarinho,  where in this case, the property is named `dim_configuration_space_` instead of `n_links` and it is defined in `DQ_Kinematics` instead of `DQ_SerialManipulator` ( I'll open a different PR to sincronize that part between the languagues).

```cpp
DQ_SerialManipulator::DQ_SerialManipulator(const int &dim_configuration_space):
    DQ_Kinematics()
{
    curr_effector_ = DQ(1);
    lower_q_limit_.resize(dim_configuration_space);
    upper_q_limit_.resize(dim_configuration_space);
    lower_q_dot_limit_.resize(dim_configuration_space);
    upper_q_dot_limit_.resize(dim_configuration_space);
    dim_configuration_space_ = dim_configuration_space;
}
```

The rationale is to define `dim_configuration_space`, 
```matlab 
obj.n_links = dim_configuration_space;
```
in one place only, instead of repeat that line of code in the concrete classes. 

Yes, we can infer the parameter during instantiation of one of the concrete subclasses, as you mentioned. But, assuming I unterstood correctly the @mmmarinho's logic, we would have to repeat the code in each of the _n_ concrete classes. For instance, if we have _n_ concrete classes, by including the `dim_configuration_space` in the constructor of _DQ_SerialManipulator_ we don't have to repeat the code _n_ times. 

<img src="https://user-images.githubusercontent.com/23158313/205825080-dac96298-c904-4f59-9579-c0a6bee599f3.png" alt="drawing" width="600"/>

What do you think?

Reply:
Hi @juanjqo,

The way you did is how I thought you'd have to do it, which contradicts the rationale you described earlier:

> The rationale is to define dim_configuration_space,
> 
> obj.n_links = dim_configuration_space;
> 
> in one place only, instead of repeat that line of code in the concrete classes. 

Having that constructor in the abstract class forces you to repeat ` obj = obj@DQ_SerialManipulator(size(A,2));` in all concrete classes. Hence the contradiction.

A simpler approach and more straightforward approach would be to do something like

```
function obj = DQ_SerialManipulatorDH(A)
obj.dim_configuration_space = size(A,2);
end
```

Both require "repeating the code in subclasses". Am I missing something?

Best wishes,
Bruno

Reply:
Hi @bvadorno, thanks for your feedback.

I assumed that each subclass required to explicitly instantiate the superclass `(obj = obj@DQ_SerialManipulator())` regardless of whether the superclass had arguments or not.  Ex:

is this exclusive to Matlab? for instance, Is it necessary in C++ to explicitly instantiate the superclass in the constructor of the subclasses? 

For instance, in matlab we have

```matlab
classdef DQ_SerialManipulatorDH < DQ_SerialManipulator
  methods
    function obj = DQ_SerialManipulatorDH(A)
           % Constructor %
    end
  end
end
```
In C++, we have
```cpp
class DQ_SerialManipulatorDH: public DQ_SerialManipulator
{
}
```

is allowed to have something like this?
```cpp
DQ_SerialManipulatorDH::DQ_SerialManipulatorDH(const MatrixXd& dh_matrix)
{
}
```
Or always it is required to instantiate the superclass? Ex:

```cpp
DQ_SerialManipulatorDH::DQ_SerialManipulatorDH(const MatrixXd& dh_matrix): DQ_SerialManipulator()
{
}
```

PS:
I implemented your suggestion

```matlab
function obj = DQ_SerialManipulatorDH(A)
obj.dim_configuration_space = size(A,2);
end
```

Best regards, 

Juancho